Add build actions

nbosb opened the pull request:
I added GitHub actions to build and test the project using the cmake and autotools workflows. The actions are triggered when new commits are pushed or a pull request is opened.

When I created the actions, I noticed some issues with the cmake workflow. I installed SystemC in a non-default location. In the CMakeLists.txt file, the `CMAKE_PREFIX_PATH` was set to `/opt/systemc`. Since I installed SystemC in another location, my SystemC version was not found. I was not able to specify my SystemC install location during configuration using `cmake -DCMAKE_PREFIX_PATH=/path/to/my/systemc/installation ...` because the value was overwritten by the CMakeLists.txt file. Therefore, I changed the CMakeLists.txt file to append the standard path `/opt/systemc` to the `CMAKE_PREFIX_PATH` variable to still enable a definition during configuration.
Another problem was that it was specified to link cci against rapidjson. Since rapidjson is a header-only library, `librapidjson` was not found and the build failed. I removed the command to link against rapidjson.

The actions build and test cci using different versions of SystemC. I think cmake support of SystemC was added in SystemC 2.3.2, so I just tested SystemC versions 2.3.2 and newer in the cmake workflow. For older SystemC versions, SystemC is not found by cmake due to missing configuration files.
For the autotools workflow, the used compiler was too new to build SystemC 2.3.0a. I got the error 

> "sorry...compiler not supported"

 when I tried to build SystemC. Therefore, I test with SystemC 2.3.1a and newer for the autotools workflow.

When SystemC 2.3.1a is used, the test `ex06_Parameter_Naming` fails with the following output:

>         SystemC 2.3.1-Accellera --- Jan 19 2023 17:28:41
>         Copyright (c) 1996-2014 by all Contributors,
>         ALL RIGHTS RESERVED
> 
> Warning: /Accellera/CCI/cci_name_gen/gen_unique_name: sim_ip.sub_ip is already used in the SystemC hierarchy, using sim_ip.sub_ip_0 instead
> In file: ../../../../../src/cci/core/cci_name_gen.cpp:82
> 
> Fatal: (F4) assertion failed: new_element && "The same parameter had been added twice!!"
> In file: ../../../../../src/cci/utils/consuming_broker.cpp:311
> FAIL ex06_Parameter_Naming/test (exit status: 134)

I am not sure if this is a bug of SystemC or cci. Either the test needs to be fixed or SystemC 2.3.1a should be removed from the tested SystemC versions.

nbosb commented:
I removed SystemC 2.3.1a from the tested versions

barnasc commented:
@nbosb : could you update the action script to select a specific ubuntu version and gcc compiler (e.g. ubuntu 22.04 with gcc11.2.0 )? Perhaps we can use this [template](https://github.com/accellera-official/cci/issues/298) indicating the implemented regressions

nbosb commented:
I reworked the matrix definitions of the actions so the OS and compiler versions that should be tested  can be specified at the top of the action file. By now, I tested

- Ubuntu 20.04 with gcc 9
- Ubuntu 20.04 with clang 9
- Ubuntu 22.04 with gcc 11
- Ubuntu 22.04 with clang 11

I added the clang test so it is easier to see how additional compilers can be added in the future. If clang tests are not wanted, I can remove them.

Currently, there are only Ubuntu runners hosted by GitHub (https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners). If other Linux distributions should be tested as well (e.g., RHEL), either a self-hosted runner needs to be used or the build action has to be restructured to run inside a docker container.

barnasc commented:
Thanks @nbosb for updating the actions. I understand that there are limitations available OSes for the runners, but perhaps we should consider MacOS and Windows since this is available as well?

aut0 commented:
This will be merged asap.
Thank you very much for your work!
